File: LeetCode/1771. Maximize Palindrome Length From Subsequences/solution.py
# class Solution:
#     def longestPalindrome(self, word1: str, word2: str) -> int:
#         n, m = len(word1), len(word2)
#         s = word1 + word2
#         dp = [[0]*(n+m+1) for _ in range(n+m+1)]
#         for i in range(n+m)[::-1]:
#             pos = {}
#             end = n+m if i >= n else n 
#             for j in range(i,end):
#                 if s[j] not in pos:
#                     pos[s[j]] = j 
#                 left, right = pos[s[j]]+1, j-1
#                 dp[i][j] = max(dp[i][j-1],dp[left][right]+2+min(right-left+1,0))
#         pos = {}
#         for i in range(n)[::-1]:
#             pos[s[i]] = i 
#             for j in range(n,n+m):
#                 left, right = pos[s[j]]+1 if s[j] in pos else -1, j-1
#                 if left == -1:
#                     if j > n:
#                         dp[i][j] = dp[i][j-1]
#                 else:
#                     dp[i][j] = max(dp[i][j-1],dp[left][right]+2) if j > n else 2
#                     if left < n:
#                         dp[i][j] = max(dp[i][j],dp[left][n-1]+2)
#                     if right >= n:
#                         dp[i][j] = max(dp[i][j],dp[n][right]+2)


#         # print(n,m,dp)
#         return dp[0][n+m-1]

# class Solution:
#     def longestPalindrome(self, word1: str, word2: str) -> int:
#         n, m = len(word1), len(word2)
#         s = word1 + word2
#         dp = [[0]*(n+m+1) for _ in range(n+m+1)]
#         for i in range(n+m)[::-1]:
#             pos = {}
#             for j in range(i,n+m):
#                 if s[j] not in pos:
#                     pos[s[j]] = j 
#                 left, right = pos[s[j]]+1, j-1
#                 if i < n and j >= n:
#                     dp[i][j] = dp[i][j-1] if j > n else 0
#                     if left <= n:
#                         dp[i][j] = max(dp[i][j],dp[left][right]+2)
#                         if left < n:
#                             dp[i][j] = max(dp[i][j],dp[left][n-1]+2)
#                         if right >= n:
#                             dp[i][j] = max(dp[i][j],dp[n][right]+2)
#                 else:
#                     dp[i][j] = max(dp[i][j-1],dp[left][right]+2+min(right-left+1,0))
       
#         # print(n,m,dp)
#         return dp[0][n+m-1]


class Solution:
    def longestPalindrome(self, word1: str, word2: str) -> int:
        n, m = len(word1), len(word2)
        s = word1 + word2
        dp = [0 for j in range(n+m)]
        dp[-1] = 1
        res = 0
        for i in range(n+m)[::-1]:   
            newdp = dp[:]
            newdp[i] = 1
            for j in range(i+1, n+m):

                if s[i] == s[j]:
                    newdp[j] = max(newdp[j], 2 + dp[j-1])
                    if i < n and j >= n:
                    #唯一的区别就在于当两端字符相等更新最长回文子串时
                    #若 i j 分别属于两个字符串，才更新最终结果值
                        res = max(res,newdp[j])
                else:
                    newdp[j] = max(dp[j], newdp[j-1])
            dp = newdp
        # print(dp)
        return res

File: LeetCode/1771. Maximize Palindrome Length From Subsequences/solution.cpp
// class Solution {
// public:
//     int longestPalindrome(string word1, string word2) {
//         int n = word1.size(), m = word2.size();
//         vector<vector<int>> dp1(n, vector<int>(n,0));
//         vector<int> left1(26, INT_MAX);
//         for (int i = n-1; i >= 0; i--){
//             dp1[i][i] = 1;
//             left1[word1[i]-'a'] = i;
//             for (int j = i+1; j < n; j++){
//                 dp1[i][j] = dp1[i][j-1];
//                 int idx = left1[word1[j]-'a'];
//                 if (idx < j){
//                     dp1[i][j] = max(dp1[i][j], dp1[idx+1][j-1]+2);
//                 }
//             }
//         }
//         vector<vector<int>> dp2(m, vector<int>(m,0));
//         vector<int> left2(26, INT_MAX);
//         for (int i = m-1; i >= 0; i--){
//             dp2[i][i] = 1;
//             left2[word2[i]-'a'] = i;
//             for (int j = i+1; j < m; j++){
//                 dp2[i][j] = dp2[i][j-1];
//                 int idx = left2[word2[j]-'a'];
//                 if (idx < j){
//                     dp2[i][j] = max(dp2[i][j], dp2[idx+1][j-1]+2);
//                 }
//             }
//         }
//         int res = 0;
//         vector<vector<int>> dp(n, vector<int>(m,0));
//         vector<int> left(26, INT_MAX);
//         for (int i = n-1; i >= 0; i--){
//             left[word1[i]-'a'] = i;
//             dp[i][0] = dp1[i][n-1];
//             int idx = left[word2[0]-'a'];
//             if (idx < INT_MAX){
//                 int len = (idx < n-1 ? dp1[idx+1][n-1] : 0) + 2;
//                 res = max(res, len);
//                 dp[i][0] = max(dp[i][0], len);
//             }
//             else{
//                 dp[i][0] = max(dp[i][0], 1);
//             }
//             for (int j = 1; j < m; j++){
//                 dp[i][j] = dp[i][j-1];
//                 int idx = left[word2[j]-'a'];
//                 if (idx < INT_MAX){
//                     int len = (idx < n-1 ? dp[idx+1][j-1] : dp2[0][j-1]) + 2;
//                     res = max(res, len);
//                     dp[i][j] = max(dp[i][j], len);
//                 }
//                 else{
//                     dp[i][j] = max(dp[i][j], dp2[0][j]);
//                 }
//             }
//         }
//         return res;
//     }
// };



class Solution {
public:
    int longestPalindrome(string word1, string word2) {
        int n = word1.size(), m = word2.size();
        auto word = word1 + word2;
        vector<vector<int>> dp(n+m, vector<int>(n+m,0));
        int res = 0;
        for (int i = n+m-1; i >= 0; i--){
            dp[i][i] = 1;
            for (int j = i+1; j < n+m; j++){
                dp[i][j] = max(dp[i][j-1], dp[i+1][j]);
                if (word[i] == word[j]){
                    int len = dp[i+1][j-1] + 2;
                    if (i < n && j >= n){
                        res = max(res, len);
                    }
                    dp[i][j] = max(dp[i][j], len);
                }
                
            }
        }
        
        return res;
    }
};